File: Scala/src/main/scala/AlgorithmicHeights/DAG.scala
package AlgorithmicHeights

object DAG {
  import scala.annotation.tailrec

  private type Component = List[Int]
  final case class Edge(a: Int, b: Int)
  final case class DFSResult(components: List[Component], previsitIds: Vector[Int], postvisitIds: Vector[Int])

  class DirectedGraph(val nrNodes: Int, edgeList: List[Edge]) {
    import DirectedGraph.buildAdjacencyList

    private val adjacencyList: Map[Int, List[Int]] = buildAdjacencyList(edgeList)

    lazy val isAcyclic: Boolean = {
      val DFSResult(_, _, postvisitIds) = depthFirstSearch()
      adjacencyList.forall{
        case (node, neighbours) =>
          val nodeNumber: Int = postvisitIds(node - 1)
          neighbours.forall(neighbour => postvisitIds(neighbour - 1) < nodeNumber)
      }
    }

    private def depthFirstSearch(): DFSResult = {
      val visitStarted: Array[Int] = Array.fill(nrNodes)(0)
      val visitEnded: Array[Int] = Array.fill(nrNodes)(0)
      val previsitId: Iterator[Int] = Iterator.from(1)
      val postvisitId: Iterator[Int] = Iterator.from(1)

      def isNodeVisited(node: Int): Boolean = visitStarted(node - 1) > 0

      def findUnvisitedNeighbour(node: Int): Option[Int] =
        adjacencyList.getOrElse(node, Nil).find(!isNodeVisited(_))

      @tailrec
      def findComponents(nodes: List[Int], components: List[Component]): List[Component] = nodes match {
        case Nil => components
        case node :: remainingNodes =>
          if (isNodeVisited(node)) findComponents(remainingNodes, components)
          else {
            val currentComponent: Component = explore(node)
            findComponents(remainingNodes, currentComponent :: components)
          }
      }

      def explore(startingNode: Int): Component = {
        @tailrec
        def traverseComponent(previsitStack: List[Int], component: Component): Component = previsitStack match {
          case Nil => component
          case node :: restOfStack =>
            findUnvisitedNeighbour(node) match {
              case Some(neighbour) =>
                visitStarted(neighbour - 1) = previsitId.next()
                traverseComponent(neighbour :: previsitStack, neighbour :: component)
              case None =>
                visitEnded(node - 1) = postvisitId.next()
                traverseComponent(restOfStack, component)
            }
        }

        visitStarted(startingNode - 1) = previsitId.next()
        traverseComponent(List(startingNode), List(startingNode))
      }

      val connectedComponents: List[Component] = findComponents((1 to nrNodes).toList, Nil)
      DFSResult(connectedComponents, visitStarted.toVector, visitEnded.toVector)
    }
  }

  object DirectedGraph {
    private def buildAdjacencyList(edgeList: List[Edge]): Map[Int, List[Int]] =
      edgeList.groupMap{ case Edge(a, _) => a }{ case Edge(_, b) => b }
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def readParameters(line: String): (Int, Int) = convertToIntList(line) match {
    case List(n, e) => (n, e)
    case _ => throw new Exception("Malformed input.")
  }

  private def readEdges(reader: Iterator[String], nrEdges: Int): List[Edge] =
    reader
      .take(nrEdges)
      .map(convertToIntList)
      .collect{ case List(a, b) => Edge(a, b) }
      .toList

  private def readGraphs(reader: Iterator[String], n: Int): List[DirectedGraph] = {
    (0 until n).map{ _ =>
      val _: String = reader.next()
      val (nrNodes, nrEdges): (Int, Int) = readParameters(reader.next())
      val edgeList: List[Edge] = readEdges(reader, nrEdges)
      new DirectedGraph(nrNodes, edgeList)
    }.toList
  }

  def testAcyclicity(graphs: List[DirectedGraph]): List[Boolean] = graphs.map(_.isAcyclic)

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val nrExamples: Int = reader.next().toInt
    val graphs: List[DirectedGraph] = readGraphs(reader, nrExamples)
    val result: List[Boolean] = testAcyclicity(graphs)
    println(result.map(if (_) "1" else "-1").mkString(" "))
  }
}

File: OCaml/src/algorithmic_heights/dag.ml
module IntMap = Map.Make(Int)

type edge = { a : int; b : int }
type dfsresult = { components : int list list; previsit_ids : int array; postvisit_ids : int array }


module Graph : sig
    type t
    val adjacency_list : t -> int list IntMap.t
    val dfs : t -> dfsresult
    val create : int -> edge list -> t
end = struct
    type t = { nr_nodes : int; adjacency_list : int list IntMap.t }

    let adjacency_list { adjacency_list; _ } = adjacency_list

    let get_neighbors adjacency_list node =
        match IntMap.find_opt node adjacency_list with
            | None -> []
            | Some neighbors -> neighbors

    let create nr_nodes edges =
        let add_edge adjacency_list { a; b } =
            let update_function = function
                | None -> Some [b]
                | Some ns -> Some (b :: ns) in
            IntMap.update a update_function adjacency_list in
        let add_edge_both_ways adjacency_list { a; b } =
            add_edge adjacency_list { a; b } in
        let adjacency_list = List.fold_left add_edge_both_ways IntMap.empty edges in
        { nr_nodes; adjacency_list }

    let dfs { nr_nodes; adjacency_list } =
        let visit_started = Array.make nr_nodes 0 in
        let visit_ended = Array.make nr_nodes 0 in
        let previsit_id = ref 1 in
        let postvisit_id = ref 1 in
        let is_node_visited node = visit_started.(node - 1) > 0 in
        let find_unvisited_neighbor node =
            node |> get_neighbors adjacency_list |> List.find_opt (fun item -> not (is_node_visited item)) in

        let explore start_node =
            let rec traverse_component component = function
                | [] -> component
                | (node :: rest) as previsit_stack ->
                    match find_unvisited_neighbor node with
                        | Some neighbor ->
                            visit_started.(neighbor - 1) <- !previsit_id;
                            previsit_id := !previsit_id + 1;
                            traverse_component (neighbor :: previsit_stack) (neighbor :: component)
                        | None ->
                            visit_ended.(node - 1) <- !postvisit_id;
                            postvisit_id := !postvisit_id + 1;
                            traverse_component rest component in
            visit_started.(start_node - 1) <- !previsit_id;
            previsit_id := !previsit_id + 1;
            traverse_component [start_node] [start_node] in

        let rec find_components components = function
            | [] -> components
            | node :: remaining_nodes ->
                if is_node_visited node then find_components components remaining_nodes
                else
                    let current_component = explore node in
                    find_components (current_component :: components) remaining_nodes in
            let components = find_components [] (List.init nr_nodes (fun k -> k + 1)) in
            { components = components; previsit_ids = visit_started; postvisit_ids = visit_ended }
end


let read_edges (nr_edges : int) : edge list =
    let parse_line line =
        Scanf.sscanf line "%d %d" (fun a b -> { a; b }) in
    let range = Seq.init nr_edges (fun k -> k) in
    List.of_seq @@ Seq.map (fun _ -> parse_line (read_line ())) range


let read_graphs (n : int) : Graph.t list =
    let read_graph () =
        ignore (read_line ());
        let (nr_nodes, nr_edges) = Scanf.sscanf (read_line ()) "%d %d" (fun n e -> (n, e)) in
        let edges = read_edges nr_edges in
        Graph.create nr_nodes edges in
    List.(map (fun _ -> read_graph ()) @@ init n (fun _ -> 0))


let is_acyclic (graph : Graph.t) : bool =
    let adjacency_list = Graph.adjacency_list graph in
    let { postvisit_ids; _ } = Graph.dfs graph in
    let check_property node neighbors =
        let node_number = postvisit_ids.(node - 1) in
        List.for_all (fun neighbor -> postvisit_ids.(neighbor - 1) < node_number) neighbors in
    IntMap.for_all check_property adjacency_list


let () =
    let nr_examples = read_int () in
    let graphs = read_graphs nr_examples in
    let results = List.map is_acyclic graphs in
    results |> List.map (fun res -> if res then "1" else "-1") |> String.concat " " |> print_endline
